include/nuttx/net/netfilter/ip6_tables.h

```diff
@@ -0,0 +1,256 @@
+#include <nuttx/net/netfilter/x_tables.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/**
+ * New IP firewall options for [gs]etsockopt at the RAW IP level.
+ * Unlike BSD Linux inherits IP options so you don't have to use
+ * a raw socket for this. Instead we check rights in the calls.
+ *
+ * ATTENTION: check linux/in6.h before adding new number here.
+ */
+
+#define IP6T_BASE_CTL                   64
+
+#define IP6T_SO_SET_REPLACE             (IP6T_BASE_CTL)
+#define IP6T_SO_SET_ADD_COUNTERS        (IP6T_BASE_CTL + 1)
+#define IP6T_SO_SET_MAX                 IP6T_SO_SET_ADD_COUNTERS
+
+#define IP6T_SO_GET_INFO                (IP6T_BASE_CTL)
+#define IP6T_SO_GET_ENTRIES             (IP6T_BASE_CTL + 1)
+#define IP6T_SO_GET_REVISION_MATCH      (IP6T_BASE_CTL + 4)
+#define IP6T_SO_GET_REVISION_TARGET     (IP6T_BASE_CTL + 5)
+#define IP6T_SO_GET_MAX                 IP6T_SO_GET_REVISION_TARGET
+
+/* Values for "inv" field in struct ip6t_ip6. */
+
+#define IP6T_INV_VIA_IN                 0x01    /* Invert the sense of IN IFACE. */
+#define IP6T_INV_VIA_OUT                0x02    /* Invert the sense of OUT IFACE */
+#define IP6T_INV_TOS                    0x04    /* Invert the sense of TOS. */
+#define IP6T_INV_SRCIP                  0x08    /* Invert the sense of SRC IP. */
+#define IP6T_INV_DSTIP                  0x10    /* Invert the sense of DST OP. */
+#define IP6T_INV_FRAG                   0x20    /* Invert the sense of FRAG. */
+#define IP6T_INV_PROTO                  XT_INV_PROTO
+#define IP6T_INV_MASK                   0x7F    /* All possible flag bits mask. */
+
+#define IP6T_STANDARD_TARGET            XT_STANDARD_TARGET
+#define IP6T_ERROR_TARGET               XT_ERROR_TARGET
+
+#define ip6t_entry_target               xt_entry_target
+#define ip6t_entry_match                xt_entry_match
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+struct ip6t_ip6
+{
+  /* Source and destination IP6 addr */
+
+  struct in6_addr src;
+  struct in6_addr dst;
+
+  /* Mask for src and dest IP6 addr */
+
+  struct in6_addr smsk;
+  struct in6_addr dmsk;
+  char iniface[IFNAMSIZ];
+  char outiface[IFNAMSIZ];
+  unsigned char iniface_mask[IFNAMSIZ];
+  unsigned char outiface_mask[IFNAMSIZ];
+
+  /* Upper protocol number
+   * - The allowed value is 0 (any) or protocol number of last parsable
+   *   header, which is 50 (ESP), 59 (No Next Header), 135 (MH), or
+   *   the non IPv6 extension headers.
+   * - The protocol numbers of IPv6 extension headers except of ESP and
+   *   MH do not match any packets.
+   * - You also need to set IP6T_FLAGS_PROTO to "flags" to check protocol.
+   */
+
+  uint16_t proto;
+
+  /* TOS to match iff flags & IP6T_F_TOS */
```

Comment:
What is this "iff flags" ?

Reply:
I think it should be if, The libip6tc.c file contains the following code:

if (e->ipv6..flags & IP6T_F_TOS)
  printf("TOS: %u\n", e->ipv6.tos);
or
if (e->ipv6.flags & IP6T_F_TOS)
  printf("%s-? %d ",
            e->ipv6.invflags & IP6T_INV_TOS ? "! " : "",
            e->ipv6.tos);
type of e->ipv6 is this struct ip6t_ip6.

Reply:
@zhhyu7 I just got this same comment in the Linux kernel source. Linux kernel uses license GPL that is not compatible with Apache. Be careful to not copy code from Linux (we want to continue to Linux-like, not a Linux license infractor). That said, in theory header files symbols are not subject to GPL according to RMS: https://linux.slashdot.org/story/11/03/20/1529238/rms-on-header-files-and-derivative-works

But never, ever, ever, look the Linux kernel source code and write a NuttX source code, otherwise you are creating derivative code :-).